XR_APILAYER_NOVENDOR_toolkit/fsr.cpp

```diff
@@ -0,0 +1,216 @@
+            m_device->setShaderInput(0, m_isSharpenOnly ? input : m_intermediary);
+            m_device->setShaderOutput(0, output, slice);
+            m_device->dispatchShader();
+        }
+
+      private:
+        void initializeScaler() {
+            const auto shadersDir = std::filesystem::path(dllHome) / std::filesystem::path("shaders");
+            const auto shaderPath = shadersDir / std::filesystem::path("FSR.hlsl");
+
+            // This value is the image region dimension that each thread group of the FSR shader operates on
+            const auto threadGroupWorkRegionDim = 16u;
+            const std::array<unsigned int, 3> threadGroups = {
+                (m_outputWidth  + (threadGroupWorkRegionDim - 1)) / threadGroupWorkRegionDim, // dispatchX
+                (m_outputHeight + (threadGroupWorkRegionDim - 1)) / threadGroupWorkRegionDim, // dispatchY
+                1 
+            };
+
+            // EASU/RCAS common
+            utilities::shader::Defines defines;
+            defines.add("FSR_THREAD_GROUP_SIZE", 64);
+            defines.add("SAMPLE_SLOW_FALLBACK", 1);
+            defines.add("SAMPLE_BILINEAR", 0);
+            defines.add("SAMPLE_HDR_OUTPUT", 0);
+
+            // EASU specific
+            defines.add("SAMPLE_RCAS", 0);
+            defines.add("SAMPLE_EASU", 1);
+            m_shaderEASU = m_device->createComputeShader(
+                shaderPath.string(), "mainCS", "FSR EASU CS", threadGroups, defines.get(), shadersDir.string());
+
+            // RCAS specific
+            defines.set("SAMPLE_EASU", 0);
+            defines.set("SAMPLE_RCAS", 1);
+            m_shaderRCAS = m_device->createComputeShader(
+                shaderPath.string(), "mainCS", "FSR RCAS CS", threadGroups, defines.get(), shadersDir.string());
+
+            m_isSharpenOnly = false;
+        }
+
+        void initializeSharpen() {
+            initializeScaler();
+
+            // TODO
+
+            //m_isSharpenOnly = true;
+        }
+
+        void initializeIntermediary(uint32_t width, uint32_t height, int64_t format) {
+            if (format == 0) {
+                // good balance between visuals and perf
+                format = m_device->getTextureFormat(TextureFormat::R16G16B16A16_UNORM);
+            }
+
+            DebugLog("FSRUpscaler initializeIntermediary with %u, %u, %u\n", width, height, format);
+
+            if (m_device->isTextureFormatSRGB(format)) {
+                format = DXGI_FORMAT_R8G8B8A8_UNORM;
+                DebugLog("  sRGB output format changed to: %u\n", format);
+            }
+
+            // create the intermediary texture between upscale and sharpen pass
+            XrSwapchainCreateInfo info;
+            ZeroMemory(&info, sizeof(info));
+            info.width = width;
+            info.height = height;
+            info.format = format;
+            info.arraySize = 1;
+            info.mipCount = 1;
+            info.sampleCount = 1;
```

Comment:
It have an issue (#14) to investigate and understand what multisampling means for our chain of textures. I think everywhere else I've just been copying this value to match the app texture. And same for `mipCount`. Maybe you can just pass the entire `info` struct to this method and just copy all the values.

Reply:
I initially wanted to pass the output texture infos along (line 102) but it felt to me like it was too much coupling between the FSR implementation, and the layer. I've therefore opted to pass texture creation parameters directly as arguments. But createTexture() is using an XrSwapchainCreateInfo argument, so I had to reconstruct one here.

Since the output texture info might not be suitable for creating the intermediary as-is, because like you said, it could have mips or what else, I suggest we keep it decoupled until further investigation about multisampling or other changes in the chain of textures.

Reply:
Yes the reuse of `XrSwapchainCreateInfo` is a bit of a shortcut of laziness - finding a common denominator between all graphics APIs. The OpenXR spec has come up with `XrSwapchainCreateInfo` to generically represent this information so I did not want to create yet-another-structure that would copy all of its content (minus `next`). A good practice might be to typedef is so it has a nicer name.

We can keep it as-is for now.